### PR TITLE
Split initial extract to store specific and combined extract file

### PR DESCRIPTION
**Pull request by NabihaMoh**
Split the initial extract file into 3 separate files: a specific extract file for scraping from ASOS, a combined extract file that identifies the website name and executes the relevant extract file and lastly, a helpers file.

### CODE REVIEW

File: .util/pylint_scores.txt
Comment (A-Faris) [inline .util/pylint_scores.txt:1]: Perfect score, nice

File: ETL/Pipeline/extract_from_asos.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+"""Extract Script: Pulls current price and sale status from ASOS API"""
+from datetime import datetime
```
Comment (A-Faris) [inline ETL/Pipeline/extract_from_asos.py]:
```suggestion
"""Extract Script: Pulls current price and sale status from ASOS API"""

from datetime import datetime
```

Comment (A-Faris) [reply ETL/Pipeline/extract_from_asos.py]: Adding empty line makes it look neater

File: ETL/Pipeline/extract_from_asos.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    """Gets the price information for a specified product from the ASOS API."""
+    if not isinstance(product_data, dict):
+        raise TypeError('product_info must be of type dict')
+    if not isinstance(headers, dict):
+        raise TypeError('header must be of type dict')
+
+    price_endpoint = get_asos_api_url(product_data['product_code'])
+
+    if not price_endpoint:
+        return None
```
Comment (A-Faris) [inline ETL/Pipeline/extract_from_asos.py]: Should there be logging error here?

File: ETL/Pipeline/extract_from_asos.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        logging.error('No valid ProductIds requested')
+        return None
+
+    return response_json[0]
+
+
+def get_current_price(product_info: dict) -> int | None:
+    """Extracts the current price of the product from the product information."""
+    if not isinstance(product_info, dict):
+        raise TypeError('product_info must be of type dict')
```
Comment (A-Faris) [inline ETL/Pipeline/extract_from_asos.py]: Any logging error here?

File: ETL/Pipeline/extract_from_asos.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        return product_info["productPrice"]["current"]["value"]
+    except KeyError as e:
+        logging.error("Error getting current price %s:", e)
+    return None
+
+
+def get_sale_status(product_info: dict) -> bool | None:
+    """Determines if the product is on sale based on its discount percentage."""
+    if not isinstance(product_info, dict):
+        raise TypeError('product_info must be of type dict')
```
Comment (A-Faris) [inline ETL/Pipeline/extract_from_asos.py]: logging error?

File: ETL/Pipeline/extract_from_asos.py
Comment (A-Faris) [inline ETL/Pipeline/extract_from_asos.py:1]: Just a few logging errors to consider

File: ETL/Pipeline/extract_combined.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    website_name = get_website_name(clean_data)
+
+    if not website_name:
+        return None
+    logging.info('Starting to run %s extract script.', website_name)
+    try:
+        website_data = EXTRACT_FUNCTIONS[website_name](product_data)
+        return website_data
+    except ValueError:
+        logging.info('extract from product %s failed! for %s ',
```
Comment (A-Faris) [inline ETL/Pipeline/extract_combined.py]: logging.error better than logging.info here

File: ETL/Pipeline/extract_combined.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    return None
+
+
+def process(product_data: dict):
+    """Processes product data by validating input, extracting the website name,
+    and running the appropriate extraction function."""
+
+    clean_data = validate_input(product_data)
+    if not clean_data:
+        return None
```
Comment (A-Faris) [inline ETL/Pipeline/extract_combined.py]: Any logging error here?

File: ETL/Pipeline/extract_combined.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    website_name = get_website_name(clean_data)
+
+    if not website_name:
+        return None
```
Comment (A-Faris) [inline ETL/Pipeline/extract_combined.py]: Any logging error here?

File: ETL/Pipeline/helpers.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
+"""This file contains functions that are used throughout this directory"""
+import logging
```
Comment (A-Faris) [inline ETL/Pipeline/helpers.py]:
```suggestion
"""This file contains functions that are used throughout this directory"""

import logging
```

File: ETL/Pipeline/helpers.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def configure_log() -> None:
+    """Configures log output"""
+    logging.basicConfig(
+        level=logging.INFO,
+        format='%(asctime)s - %(levelname)s - %(message)s'
```
Comment (A-Faris) [inline ETL/Pipeline/helpers.py:10]: Nice formatting

File: ETL/Pipeline/helpers.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    )
+
+
+def has_required_keys(entry, required_keys):
+    """Check if all required keys are present in the dictionary."""
+    return all(key in entry for key in required_keys)
```
Comment (A-Faris) [inline ETL/Pipeline/helpers.py:16]: I love the efficency here

File: ETL/Pipeline/helpers.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def is_dict(entry):
+    """Check if the entry is a dictionary."""
+    return isinstance(entry, dict)
```
Comment (A-Faris) [inline ETL/Pipeline/helpers.py]: Is this function actually needed? Feels too small and it's only used once

File: ETL/Pipeline/helpers.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    """Validates the input list of products"""
+    required_keys = {
+        'product_id': int,
+        'url': str,
+        'product_code': int,
+        'product_name': str
+    }
+
+    if (is_dict(entry) and has_required_keys(entry, required_keys) and
+            convert_product_code(entry) and has_correct_types(entry, required_keys)):
```
Comment (A-Faris) [inline ETL/Pipeline/helpers.py]:
```suggestion
    if isinstance(entry, dict) and has_required_keys(entry, required_keys) and
            convert_product_code(entry) and has_correct_types(entry, required_keys):
```

File: ETL/Pipeline/helpers.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+        return entry
+    logging.error("Entry is NOT valid")
+
+    return None
```
Comment (A-Faris) [inline ETL/Pipeline/helpers.py:49]: Either no space or put space there. I'm being very picky here
```suggestion

    logging.error("Entry is NOT valid")
    return None
```

File: ETL/Pipeline/helpers.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def remove_stale_products(products: list[dict]) -> list[dict]:
+    """remove products where the price hasn't decreased!"""
+    return [product for product in products
+            if product.get('previous_price') is None
+            or product['current_price'] <= product['previous_price']]
```
Comment (A-Faris) [inline ETL/Pipeline/helpers.py:56]: I love the efficiency here

File: ETL/Pipeline/helpers.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+               for key, required_type in required_keys.items())
+
+
+def convert_product_code(entry):
+    """Convert the product_code from str to int."""
+    try:
+        entry['product_code'] = int(entry['product_code'])
+        return True
+    except (ValueError, TypeError):
+        return False
```
Comment (A-Faris) [inline ETL/Pipeline/helpers.py]: logging error?

File: ETL/Pipeline/extract_combined.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+def extract_price_and_sales_data(product_list: list[dict]) -> list[dict]:
+    """Populates each product dictionary in the product list with current price, reading time,
+    and sale status using multiprocessing."""
+    logging.info("Starting Extraction")
+    with Pool(processes=4) as pool:
+        logging.info("Adding the current price and sale status")
+        results = list(pool.map(process, product_list))
+        logging.info("Finished Extraction")
+        logging.info("Removing erroneous / missing data")
```
Comment (A-Faris) [inline ETL/Pipeline/extract_combined.py]: Consider combining them

**Review by A-Faris — CHANGES_REQUESTED** [review]
A few logging errors and formatting to consider

**Review by BerkayDur — DISMISSED** [review]
I looks good.

File: ETL/Pipeline/extract_combined.py [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+            'product_id': 3,
+            'product_code': 205918844,
+            'url': "https://www.asos.com/pasq/pasq-two-pocket-tote-bag-\
+        with-removable-pouch-in-black/prd/205928631#colourWayId-205928635",
+            'product_name': 'PASQ two pocket tote bag with removable pouch in black',
+            'website_name': 'asos'
+        }
+    ]
+
+    print(extract_price_and_sales_data(inputs))
```
Comment (A-Faris) [inline ETL/Pipeline/extract_combined.py]: Destroy this please

**Review by A-Faris — DISMISSED** [review]
One little change but amazing otherwise

**Review by A-Faris — APPROVED** [review]
Good work

**Pull request merged**